injector.c

```diff
@@ -57,10 +59,12 @@ void injector_close(void) {
     LATB5 = 0;
     LATB2 = 0;
     LATB3 = 0;
-
-    __delay_us(200);
-    LATB4 = 1;  // high side A -> high
-    LATB3 = 1;  // low side B -> high
+    if(injector_open_called_last) {
+        injector_open_called_last = false;
+        __delay_us(200);
+        LATB4 = 1;  // high side A -> high
+        LATB3 = 1;  // low side B -> high
+    }
```

Comment:
I see what you're going for here, but your reasoning is actually backwards. Right now, if `injector_close` is called twice (the first time being after an open), the first "close" will  actually close the valve, while the second one will cut off all power. This is a problem for a few reasons:

1. The function leaves the valve in a different state depending on when it's called. One case leaves the "close" pins on, while the other leaves everything cut off. This is generally considered poor form. You want the function to have the same behaviour each time.

2. Calling this function twice in rapid succession leaves the valve in a bad state. The first time you call it, the `close` pins are activated and the valve begins to move. The second time you call it, you cut off all power, stopping it from moving. If not much time has passed between the two calls, the valve is now stopped somewhere between states. This is not a good thing (eg. we accidentally only open the injector a little bit, which ruins our launch).

3. The way this function is written breaks some of the error checks in error_checks.c. That check makes sure that when we want the valve to be open, we're actually telling the appropriate pins to do that. This violates that since we now cut off all the pins.

You do avoid the delay statement, but that's not really the problem with this function the way it was before. The problem was that if we called the function twice in rapid succession, it would briefly stop moving the valve before continuing. If we called it many times in rapid succession, it would effectively PWM the mosfets, which we want to avoid.

The proper way to write this function would be something like this (pseudocode):
```
injector_close():
    if current_state == VALVE_CLOSED:
        // everything is fine, don't do anything
        return
    else:
        // close injector
        LATx = 0;   // not valid code lol, just turn all the pins off
        delay()      // delay a bit to avoid shoot-through through the bridge
        injector_close_pins_on
        current_state = VALVE_OPEN
```

Reply:
So for the case where injector_close() is called again (even thought the injector is already closed), should anything happen? More specifically, should this be contained within the if statement or within the else statement:
`WHITE_LED_OFF();
        // turn everything off for a bit to avoid shoot-through
        LATB4 = 0;
        LATB5 = 0;
        LATB2 = 0;
        LATB3 = 0;`

Reply:
In the else, as shown in the pseudocode above ("LATx = 0"). Otherwise you are cutting power to the valve if it is already "closed" which is a bad thing for the reasons outlined above.

Reply:
That makes more sense. Thank you!